### DISCUSSION TITLE
Conditional formatting cell (Color and Background Color) with custom plugin

### DISCUSSION THREAD

**Discussion by asananddevsingh** [handsontable/hyperformula#1342] [API Questions]
Hi, 

I am working on a feature POC where I want to allow the user to insert a column to the handsontable and fill that with data from any existing column with some conditional formatting. 

Handontable used with "non-commercial-and-evaluation" license with below versions
- "handsontable": "13.1.0",
- "@handsontable/react": "13.1.0",
- "hyperformula": "2.6.0"

I've created a custom plugin for hyperformula called as "FORMAT" whick takes 3 arguments, 
e.g. =FORMAT(B1 > 100, "white", "red") 

1. Condition
2. Cell Text color
3. Cell Background color

Now, I want to parse this formaula and format the cell based on it. I am unabele to understand how to control the Handsontable  from Plugin. Grid example is as below:

[Here ](https://codesandbox.io/p/sandbox/hansontable-poc-forked-c5gy7r?file=%2Fsrc%2Fhandsontable%2FCustomPlugin.js&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clpcpj28h0007356g6z6tea4b%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clpcpj28h0003356ggk15t60t%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clpcpj28h0004356g4e15jw7y%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clpcpj28h0006356gwr7maz4r%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B55.688429217840984%252C44.311570782159016%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clpcpj28h0003356ggk15t60t%2522%253A%257B%2522id%2522%253A%2522clpcpj28h0003356ggk15t60t%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpcrozk90003356gu5rnzbd6%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A33%252C%2522startColumn%2522%253A20%252C%2522endLineNumber%2522%253A33%252C%2522endColumn%2522%253A20%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Fhandsontable%252FDataSheet.jsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clpcs90aj0003356gttdwt7bg%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A16%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A16%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Fhandsontable%252FCustomPlugin.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clpcs90aj0003356gttdwt7bg%2522%257D%252C%2522clpcpj28h0006356gwr7maz4r%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpcpj28h0005356gni8hllu9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%2522%257D%255D%252C%2522id%2522%253A%2522clpcpj28h0006356gwr7maz4r%2522%252C%2522activeTabId%2522%253A%2522clpcpj28h0005356gni8hllu9%2522%257D%252C%2522clpcpj28h0004356g4e15jw7y%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clpcpj28h0004356g4e15jw7y%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) is Sandbox of the my code:

e.g. 
![image](https://github.com/handsontable/hyperformula/assets/13918558/7973c266-c748-47aa-9ed1-d6f07e268ec0)

Please suggest, if it is possible to allow user to add column and apply formatting formula to it.

Thanks, 
Anand Dev

**Comment by thilgen** [comment]
@asananddevsingh - I don't think you will be able to do this with a custom HyperFormula function plugin. HyperFormula is "headless" and doesn't know anything about how a host application (like Handsontable) elects to render/draw a cell or its contents. 

If you are interested in getting into some internals - I worked on a feature earlier this year that allowed HyperFormula to pass a "hint" to the host application that a cell should be rendered as an HTML Link that you might find useful. The HyperFormula part has been accepted but the Handsontable code is still pending.

Issue: https://github.com/handsontable/hyperformula/issues/1215
HyperFormula: https://github.com/handsontable/hyperformula/pull/1216 (merged)
Handsontable: https://github.com/handsontable/handsontable/pull/10314 (open)

**Note** You can see a bit of this in your sandbox/sample if you add a `console.log(ast)` - the `hyperlink` property on the `ast` object is the "hint" passed to the host application that the cell contains an HTML link and that (if the host application chooses) it can be rendered with a different color and link-click semantics.

**Reply by asananddevsingh** [reply]
Hi @thilgen, 

Thank you for swift response. I appreciate your effort.

I've looked at the "Note" you suggested about "ast" and I can see its value as `undefined`. Now, could you please help me with --From where can I set the value to "hyperlink" flag? or how to set to set its value as a string as hit for my handsontable?

Also, is there any way available with Handsontable (without using custom plugin/hyperformula) by using which I can give the user an ability to insert column dynamically and set some color based on some condition. 

Something  like this example but conditionally: https://jsfiddle.net/d1ye98u7/ 

I am ready to buy the license too, if it's available in pro version.

![image](https://github.com/handsontable/hyperformula/assets/13918558/39a922f5-1c91-4e8b-9e1a-e1cdf2778014)

**Comment by AMBudnik** [comment]
Hi @asananddevsingh 

Is there any particular need to have a custom formula? If the cell color should change based on the value of the 2nd column you can use IF and some cell metadata conditionals.

Here's an example https://jsfiddle.net/9wpamzxu/

I set it up to show nothing if the condition is not met.

**Edition**
![Zrzut ekranu 2023-11-28 o 14 31 48](https://github.com/handsontable/hyperformula/assets/10757813/316dcb58-f9e3-4396-98cf-e8e862cf7eee)


**Outcome** 
![Zrzut ekranu 2023-11-28 o 14 31 56](https://github.com/handsontable/hyperformula/assets/10757813/0988e8ad-dda2-4b66-9d32-4f058c5ea44a)

**Reply by asananddevsingh** [reply]
Hi @AMBudnik,

Thank you for reply. I appreciate it. 

The example you shared is exactly what I want but the catch is, --I want the user to decide to insert column at any index, that means I cannot check the column index in the code.

![image](https://github.com/handsontable/hyperformula/assets/13918558/4ba38804-1e23-4931-88f9-074cbd180a18)

Is there any way to keep it index independent or any way to allow user to insert special type of column. That I can know in code as rendered column. 

Also, If I am using the HotColumn then the ability to "insert column" goes away. Even when I keep allowInsertColumn to true. I don't know why it happens.

**Reply by AMBudnik** [reply]
Would you use formulas in any other column? If so, you can use that like this https://jsfiddle.net/dqjsuab4/ (triggered only when a cell source starts with `=`).

**Reply by asananddevsingh** [reply]
Yes, it would be having other columns too with formulas to calculate the values.

**Reply by AMBudnik** [reply]
OK, so you're right - the custom function might be needed to extend that check from the line `13` from `=` to `=FORMAT`.

**Reply by asananddevsingh** [reply]
Thanks for response @AMBudnik. 

I think the same way. Is there any callback or hook available to refresh the cell/table once custom plugin updates the cell data?

I am planning to get the details from user and set everything as string in cell and then parse that string and refresh the data in cell at handsontable level (React component/rendered level).

![image](https://github.com/handsontable/hyperformula/assets/13918558/6a206c34-197d-47b6-8fda-100cb6b3fe90)

Also, is there any way to use HotColumn component (in react) and also allow the user to insert column?

**Reply by AMBudnik** [reply]
There's a hook called `afterSetCellMeta()` it is triggered once we use `setCellMeta()` method. Now refreshing cells could be done via `render()` method. you need it if you want the changes to apply visually to the table once you use `setCellMeta()`.

**Reply by asananddevsingh** [reply]
I've tried these methods but no luck. Thanks for replies. I apprceate it.

**Comment by sequba** [comment]
In my opinion, conditional formatting should be the responsibility of the UI (hansontable or a custom one), so HyperFormula does not need any changes to make it possible.

I believe this discussion is more about [handsontable](https://github.com/handsontable/handsontable) API now.

**Reply by AMBudnik** [reply]
Yes, @sequba you're right.

Can you update me on the latest changes at support@handsontable.com? We'd like to close the issue here as it is not strictly Hyperformula-related

**Reply by asananddevsingh** [reply]
Yes @AMBudnik , please close this. 

I am dealing with this need using Handsontable and React Js ways.

I appreciate your time and effort.
